ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,767 @@
+	}
+
+	DMSG("Create master key secret");
+
+	TEE_GenerateRandom(secret_data, sizeof(secret_data));
+
+	res = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+			secret_id,
+			sizeof(secret_id),
+			TEE_DATA_FLAG_ACCESS_WRITE,
+			TEE_HANDLE_NULL, NULL, 0,
+			&secret_obj);
+	if (res) {
+		EMSG("Failed to create secret");
+		return res;
+	}
+
+	res = TEE_WriteObjectData(secret_obj, (void *) secret_data,
+			sizeof(secret_data));
+	if (res)
+		EMSG("Failed to write secret data");
+
+exit:
+	if (secret_obj != TEE_HANDLE_NULL)
+		TEE_CloseObject(secret_obj);
+
+	return res;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS] __unused,
+				    void **sess_ctx __unused)
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	init_failure_record();
+
+	return TEE_SUCCESS;
+}
+
+void TA_CloseSessionEntryPoint(void *sess_ctx __unused)
+{
+}
+
+static TEE_Result TA_GetMasterKey(TEE_ObjectHandle master_key)
+{
+	TEE_Result res;
+	TEE_Attribute attr;
+	uint8_t	secret_data[TEE_SHA256_HASH_SIZE];
+	TEE_ObjectHandle secret_obj = TEE_HANDLE_NULL;
+	uint32_t read_size = 0;
+
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE, secret_id,
+				       sizeof(secret_id),
+				       TEE_DATA_FLAG_ACCESS_READ,
+				       &secret_obj);
+	if (res) {
+		EMSG("Failed to open secret, error=%X", res);
+		goto exit;
+	}
+
+	res = TEE_ReadObjectData(secret_obj, secret_data, sizeof(secret_data),
+				 &read_size);
+	if (res || sizeof(secret_data) != read_size) {
+		EMSG("Failed to read secret data, bytes = %u", read_size);
+		goto close_obj;
+	}
+
+	TEE_InitRefAttribute(&attr, TEE_ATTR_SECRET_VALUE, secret_data,
+			     sizeof(secret_data));
+
+	res = TEE_PopulateTransientObject(master_key, &attr, 1);
+	if (res)
+		EMSG("Failed to set master key attributes");
+
+close_obj:
+	TEE_CloseObject(secret_obj);
+exit:
+	return res;
+}
+
+static TEE_Result TA_ComputeSignature(uint8_t *signature,
+				      size_t signature_length,
+				      TEE_ObjectHandle key,
+				      const uint8_t *message,
+				      size_t length)
+{
+	uint32_t buf_length = TEE_SHA256_HASH_SIZE;
+	uint8_t buf[buf_length];
+	TEE_OperationHandle op = TEE_HANDLE_NULL;
+	TEE_Result res;
+	uint32_t to_write;
+
+	res = TEE_AllocateOperation(&op, TEE_ALG_HMAC_SHA256, TEE_MODE_MAC,
+				    TEE_SHA256_HASH_SIZE * 8);
+	if (res) {
+		EMSG("Failed to allocate HMAC operation");
+		goto exit;
+	}
+
+	res = TEE_SetOperationKey(op, key);
+	if (res) {
+		EMSG("Failed to set secret key");
+		goto free_op;
+	}
+
+	TEE_MACInit(op, NULL, 0);
+
+	TEE_MACComputeFinal(op, (void *)message, length, buf, &buf_length);
+	if (res) {
+		EMSG("Failed to compute HMAC");
+		goto free_op;
+	}
+
+	to_write = buf_length;
+
+	if (buf_length > signature_length)
```

Comment:
Why should these even differ? Isn't it always a HMAC-SHA256? I.e., a well known signature size? I.e., a check in the start of the function instead like ...
```
if (signature_length != TEE_SHA256_HASH_SIZE)
   return TEE....bla_bla_bla;
```
sounds more correct, or?